src/parser.cpp

```diff
@@ -913,7 +922,7 @@ namespace ada::parser {
             }
 
             // Let host be the result of host parsing buffer with url is not special.
-            std::optional<ada::url_host> host = parse_host(buffer, !url.is_special());
+            std::optional<ada::url_host> host = parse_host(buffer, !url.is_special(), is_ascii);
```

Comment:
While adding to buffer, if we also checked for `if (unicode::is_forbidden_domain_code_point(c)) {` we can remove the unnecessary iteration on `parse_host` function

Reply:
Ok.

Reply:
I have implemented this idea of yours. See my last commit.